src/rdmap.c

```diff
@@ -0,0 +1,487 @@
+                buckets.cnt = 199;
+        } else {
+                /* Strive for a maximum depth of 15 elements per bucket, but
+                 * limit the maximum bucket size.
+                 * When a real need arise we'll change this to a dynamically
+                 * growing hash map instead, but this will do for now. */
+                buckets.cnt = bucket_sizes[0];
+                for (i = 1 ; bucket_sizes[i] != -1 &&
+                             (int)expected_cnt / max_depth > bucket_sizes[i];
+                     i++)
+                        buckets.cnt = bucket_sizes[i];
+        }
+
+        rd_assert(buckets.cnt > 0);
+
+        buckets.p = rd_calloc(buckets.cnt, sizeof(*buckets.p));
+
+        return buckets;
+}
+
+
+void rd_map_init (rd_map_t *rmap, size_t expected_cnt,
+                  int (*cmp) (const void *a, const void *b),
+                  unsigned int (*hash) (const void *key),
+                  void (*destroy_key) (void *key),
+                  void (*destroy_value) (void *value)) {
+
+        memset(rmap, 0, sizeof(*rmap));
+        rmap->rmap_buckets = rd_map_alloc_buckets(expected_cnt);
+        rmap->rmap_cmp = cmp;
+        rmap->rmap_hash = hash;
+        rmap->rmap_destroy_key = destroy_key;
+        rmap->rmap_destroy_value = destroy_value;
+}
+
+void rd_map_destroy (rd_map_t *rmap) {
+        rd_map_elem_t *elem;
+
+        while ((elem = LIST_FIRST(&rmap->rmap_iter)))
+                rd_map_elem_destroy(rmap, elem);
+
+        rd_free(rmap->rmap_buckets.p);
+}
+
+
+int rd_map_str_cmp (const void *a, const void *b) {
+        return strcmp((const char *)a, (const char *)b);
+}
+
+/**
+ * @brief A djb2 string hasher.
+ */
+unsigned int rd_map_str_hash (const void *key) {
+        const char *str = key;
+        return rd_string_hash(str, -1);
+}
+
+
+
+/**
+ * @name Unit tests
+ *
+ */
+#include "rdtime.h"
+#include "rdunittest.h"
+#include "rdcrc32.h"
+
+
+/**
+ * Typed hash maps
+ */
+
+/* Complex key type */
+struct mykey {
+        int k;
+        int something_else;  /* Ignored by comparator and hasher below */
+};
+
+/* Key comparator */
+static int mykey_cmp (const void *_a, const void *_b) {
+        const struct mykey *a = _a, *b = _b;
+        return a->k - b->k;
+}
+
+/* Key hasher */
+static unsigned int mykey_hash (const void *_key) {
+        const struct mykey *key = _key;
+        return (unsigned int)key->k;
+}
+
+/* Complex value type */
+struct person {
+        char *name;
+        char *surname;
+};
+
+/* Define typed hash map type */
+typedef RD_MAP_TYPE(const struct mykey *, const struct person *)
+        ut_my_typed_map_t;
+
+
+/**
+ * @brief Test typed hash map with pre-defined type.
+ */
+static int unittest_typed_map (void) {
+        ut_my_typed_map_t rmap = RD_MAP_INITIALIZER(0,
+                                                    mykey_cmp, mykey_hash,
+                                                    NULL, NULL);
+        struct mykey k1 = { 1 };
```

Comment:
maybe better to explicitly specify a value for `something_else`. side note: c language seems generally screwed up to me 
on struct initialization.

Reply:
Since something_else isn't used the auto initialization to zero should be fine.
What's the screwy part?